### PR TITLE
WIP Fix akka-http test instrumentation

### PR DESCRIPTION
**Pull request by devinsba**
I created this as part of my `bindAndHandle` branch but wanted to get it tested against master because this is valuable outside of `bindAndHandle`

### CODE REVIEW

File: dd-java-agent/instrumentation/play-2.6/src/test/groovy/server/PlayHttpTestInstrumentation.java [hunk 1]
```diff
@@ -6,7 +8,7 @@
 import net.bytebuddy.agent.builder.AgentBuilder;
 
 @AutoService(Instrumenter.class)
-public class AkkaHttpTestInstrumentation implements Instrumenter {
+public class PlayHttpTestInstrumentation implements Instrumenter {
```
Comment (devinsba) [inline dd-java-agent/instrumentation/play-2.6/src/test/groovy/server/PlayHttpTestInstrumentation.java]: I'll copy the above down to here assuming it is :+1:'d

Comment (tylerbenson) [reply dd-java-agent/instrumentation/play-2.6/src/test/groovy/server/PlayHttpTestInstrumentation.java]: seems ok to me assuming the tests pass and it's verifying the `ignoreParent` behavior.

File: dd-java-agent/instrumentation/akka-http-10.0/akka-http-10.0.gradle [hunk 2]
```diff
@@ -101,8 +101,8 @@ dependencies {
   latestDepTestCompile project(':dd-java-agent:instrumentation:trace-annotation')
 }
 
-test.dependsOn lagomTest
-test.dependsOn version101Test
+check.dependsOn lagomTest
+check.dependsOn version101Test
```
Comment (devinsba) [inline dd-java-agent/instrumentation/akka-http-10.0/akka-http-10.0.gradle]: Need to revert

File: dd-java-agent/instrumentation/java-concurrent/src/main/java/datadog/trace/instrumentation/java/concurrent/AdviceUtils.java [hunk 3]
```diff
@@ -24,6 +24,7 @@
       final TraceScope.Continuation continuation = state.getAndResetContinuation();
       if (continuation != null) {
         final TraceScope scope = continuation.activate();
+        log.debug("startTaskScope() {} {}", task, scope);
```
Comment (devinsba) [inline dd-java-agent/instrumentation/java-concurrent/src/main/java/datadog/trace/instrumentation/java/concurrent/AdviceUtils.java]: revert

File: dd-java-agent/instrumentation/java-concurrent/src/main/java/datadog/trace/instrumentation/java/concurrent/ExecutorInstrumentationUtils.java [hunk 4]
```diff
@@ -44,10 +44,16 @@ public static boolean shouldAttachStateToTask(final Object task, final Executor
    */
   public static <T> State setupState(
       final ContextStore<T, State> contextStore, final T task, final TraceScope scope) {
+    log.debug("stacktrace", new Exception(task.getClass().getName()));
```
Comment (devinsba) [inline dd-java-agent/instrumentation/java-concurrent/src/main/java/datadog/trace/instrumentation/java/concurrent/ExecutorInstrumentationUtils.java]: revert

File: dd-trace-ot/src/main/java/datadog/opentracing/DDSpan.java [hunk 5]
```diff
@@ -83,6 +83,7 @@ public boolean isFinished() {
   }
 
   private void finishAndAddToTrace(final long durationNano) {
+    log.debug("stacktrace", new Exception());
```
Comment (devinsba) [inline dd-trace-ot/src/main/java/datadog/opentracing/DDSpan.java]: revert

File: dd-trace-ot/src/main/java/datadog/opentracing/PendingTrace.java [hunk 6]
```diff
@@ -226,6 +226,9 @@ private void expireReference() {
       }
     }
     log.debug("traceId: {} -- Expired reference. count = {}", traceId, count);
+    for (final WeakReference reference : weakReferences) {
+      log.debug("Reference {}", reference.get());
+    }
```
Comment (devinsba) [inline dd-trace-ot/src/main/java/datadog/opentracing/PendingTrace.java]: revert

File: dd-trace-ot/src/main/java/datadog/opentracing/scopemanager/ContinuableScope.java [hunk 7]
```diff
@@ -58,6 +58,7 @@ private ContinuableScope(
```
Comment (devinsba) [inline dd-trace-ot/src/main/java/datadog/opentracing/scopemanager/ContinuableScope.java]: revert all in this file

File: dd-java-agent/instrumentation/java-concurrent/src/main/java/datadog/trace/instrumentation/java/concurrent/AkkaForkJoinTaskInstrumentation.java [hunk 8]
```diff
@@ -114,6 +115,10 @@ public static TraceScope enter(@Advice.This final ForkJoinTask thiz) {
 
     @Advice.OnMethodExit(onThrowable = Throwable.class, suppress = Throwable.class)
     public static void exit(@Advice.Enter final TraceScope scope) {
+      // Clean up dandling scopes from async code
+      while (activeScope() != scope && activeScope() != null) {
+        activeScope().close();
+      }
```
Comment (devinsba) [inline dd-java-agent/instrumentation/java-concurrent/src/main/java/datadog/trace/instrumentation/java/concurrent/AkkaForkJoinTaskInstrumentation.java]: @DataDog/apm-java this is ultimately what fixes the dangling scope created in the test wrapper. This only works for akka-http 10.1, it seems that 10.0 fork join tasks don't finish before they create new continuations for that scope.

I almost feel like activating a continuation that has a scope for a closed span should close the scope but I feel like this is now starting to look like potential bugs.

Ultimately I think the long term fix should be a more targeted set of akka fork-join/executor instrumentations than the general java ones

Comment (tylerbenson) [reply dd-java-agent/instrumentation/java-concurrent/src/main/java/datadog/trace/instrumentation/java/concurrent/AkkaForkJoinTaskInstrumentation.java]: I really don't think we should merge this change.  This is a pretty scary work around

Comment (devinsba) [reply dd-java-agent/instrumentation/java-concurrent/src/main/java/datadog/trace/instrumentation/java/concurrent/AkkaForkJoinTaskInstrumentation.java]: 100% agree. I changed the title to WIP to make sure that was obvious. I'm at a loss for fixing this without some major renovation to the akka instrumentations. We may need to instrument akka-streams directly first to deal with the flows

File: dd-java-agent/instrumentation/akka-http-10.0/src/test/groovy/AkkaHttpTestInstrumentation.java [hunk 9]
```diff
@@ -1,19 +1,163 @@
...
+                public void onPush() throws Exception {
+                  final HttpRequest request = grab(requestInlet);
+
+                  HttpServerTestAdvice.ServerEntryAdvice.methodEnter();
+                  final TraceScope traceScope = activeScope();
+                  if (traceScope != null) {
+                    agentScopes.add(traceScope.capture());
+                  }
+
+                  push(requestOutlet, request);
```
Comment (devinsba) [inline dd-java-agent/instrumentation/akka-http-10.0/src/test/groovy/AkkaHttpTestInstrumentation.java]: @tylerbenson I've been pondering this all morning. I think the right solution to fixing this scoping issue is to instrument akka-streams in such a way that this call can be made with an active scope, and then close it before exiting this method.

So in essence I don't think we can fix this without instrumenting akka-streams

Comment (tylerbenson) [reply dd-java-agent/instrumentation/akka-http-10.0/src/test/groovy/AkkaHttpTestInstrumentation.java]: That sounds like a reasonable conclusion.  We had to do something similar with rx-java.

**Pull request closed without merging**